src/main/java/com/shapesecurity/functional/data/NonEmptyImmutableList.java

```diff
@@ -42,6 +43,14 @@ protected NonEmptyImmutableList(@NotNull T head, @NotNull final ImmutableList<T>
         return tail;
     }
 
+    @NotNull
+    private T[] toObjectArray() {
```

Comment:
I'm not so sure. This is actually not a `T[]` but an `Object[]` at runtime. I did it here by the means of "[heap pollution](https://docs.oracle.com/javase/tutorial/java/generics/nonReifiableVarargsType.html#heap_pollution)". It means if you pass it around, it type checker of JVM will fail at run time.

One example is that you cannot downcast an `Object[]` to an `T[]` unless `T` is actually `Object`:
```java
Integer[] array = ImmutableList.from(1,2,3).toArray(); // ClassCastException at runtime.
```
There is no explicit cast but the implisit cast from `Integer(as a type argument)` to `Integer` happened due to type erasure.

It is fine to use if you only use then in the context of a type parameter (which's why I can cast it to `T[]` in the first place).